### DISCUSSION TITLE
Is there any way to get the name of a field as a string?

### DISCUSSION THREAD

**Discussion by rohan-mehta** [pydantic/pydantic#8600] [Question] [answered]
Let's say I have this model:
```py
class Person(BaseModel):
  first_name: str
```

I want to get the literal string `"first_name"`. Maybe via something like `Person.first_name.__field_name__`.

## Use case:
I store JSON representations of the models in a nosql DB. But if I want to filter my queries, I need to hardcode strings e.g.:
```py
db.where("first_name", "==", "Rohan")
```
But this is fragile since the field name could change and my typechecker has no idea. And queries would silently fail. Much nicer would be:
```py
db.where(Person.first_name.__field_name__, "==", "Rohan")
```

## Things I've considered
`Person.model_fields["first_name"]`. This is still hardcoding, but at least it's a runtime error (as opposed to silent failure)

Please let me know if there's a good way to achieve this! Thanks :)

**Comment by uriyyo** [comment, marked as answer]
Hi @rohan-mehta,

I guess there is no built-in way to do such things, but you can introduce helper class:
```py
from dataclasses import dataclass
from functools import lru_cache
from typing import TypeVar, cast, Any, reveal_type, TYPE_CHECKING

from pydantic import BaseModel


class Person(BaseModel):
    first_name: str


@dataclass(frozen=True)
class _GetFields:
    _model: type[BaseModel]

    def __getattr__(self, item: str) -> Any:
        if item in self._model.model_fields:
            return item

        return getattr(self._model, item)


TModel = TypeVar("TModel", bound=type[BaseModel])


def fields(model: TModel, /) -> TModel:
    return cast(TModel, _GetFields(model))


if not TYPE_CHECKING:
    fields = lru_cache(maxsize=256)(fields)

person_fields = fields(Person)
reveal_type(person_fields)

first_name = fields(Person).first_name
reveal_type(first_name)

print(first_name)
```
```py
note: Revealed type is "def (*, first_name: Any, **kwargs: Any) -> biz.Person"
note: Revealed type is "builtins.str"
```
```py
first_name
```

Usage will look like this:
```py
db.where(fields(Person).first_name, "==", "Rohan")
```

**Reply by rohan-mehta** [reply]
Thanks, this is a pretty solid workaround. Will try it!

**Reply by samuelkolb** [reply]
Really nice solution!  Small comment: To get correct auto completion I think you should use

```py
TModel = TypeVar("TModel", bound=BaseModel)


def fields(model: type[TModel], /) -> TModel:
    return cast(TModel, _GetFields(model))
```

(The input is a type, but you would call the fields on an instance)

**Reply by wdhwg001** [reply]
In static type checking, the `first_name` var is `str` is because the `first_name` field in `Person` is a `str`, but suppose we have:

```python3
class Person(BaseModel):
    first_name: str
    age: int
```

Then:

```python3
age = fields(Person).age
reveal_type(age)
```

While in the runtime it gives: `Runtime type is 'str'`, but Pylance said it's `int`:

![image](https://github.com/user-attachments/assets/021df5e2-fee3-47a6-929c-6deb8215823a)

**Reply by danyiimp** [reply]
An ugly but working solution with double casting:
```python
def fc(field: Any):
    """ Pydantic model field cast """
    return cast(str, field)

age = fc(fields(Person).age)
reveal_type(age)
```

```
Type of "age" is "str" Pylance
```

**Reply by wdhwg001** [reply]
That's essentially the same as `str(fields(Person).age)`.

**Comment by Girmii** [comment]
For anyone coming across this, this is how I've solved this myself;

```python
from pydantic import BaseModel
from pydantic._internal._model_construction import ModelMetaclass


class _BaseModelGetAttrMeta(ModelMetaclass):
    def __getattr__(self, item: str):
        try:
            super().__getattr__(item)
        except AttributeError:
            if item in self.__dict__.get("__pydantic_fields__", ()):
                return item
            else:
                raise


class BaseModelGetAttr(BaseModel, metaclass=_BaseModelGetAttrMeta):
    pass

class MyModel(BaseModelGetAttr):
    name: str
    id: int


print(f"{MyModel.name=}")  # MyModel.name='name'
print(f"{MyModel.id=}")    # MyModel.id='id'
```

Not sure if there is a more concise way to achieve this, or whether this breaks any other functionality, but so far it works for my purposes.

See also this issue; https://github.com/pydantic/pydantic/discussions/8399
